### PR TITLE
Implement DTXProfiler subset for Android

### PR DESCRIPTION
**Pull request by valentynberehovyi**
Added basic integration with DetoxInstruments for Android.

### CODE REVIEW

File: detox/src/artifacts/instruments/recording/AndroidRecordingPathCreator.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
+const RecordingPathBuilder = require('./RecordingPathCreator');
+
+class AndroidRecordingPathCreator extends RecordingPathBuilder {
+  constructor(devicePathBuilder) {
+    super();
+    this.devicePathBuilder = devicePathBuilder;
+  }
+
+  createRecordingPath() {
+    return this.devicePathBuilder.buildTemporaryArtifactPath('.dtxplain');
```
[the remark this answers is not in the document]
Comment (valentynberehovyi) [reply detox/src/artifacts/instruments/recording/AndroidRecordingPathCreator.js]: RecordingPathCreator is abstraction which may have different implementations as example for android and iOS when DevicePathBuilder is just an android util.

File: detox/test/android/app/src/main/java/com/example/NativeModule.java [hunk 2]
```diff
@@ -111,42 +110,40 @@ public void getLaunchArguments(Promise promise) {
...
-     * called concurrently in the ActionsScreen of this demo app; could it be that it is not always called
-     * from the main thread?).
+     * If real life (CI), we've genuinely seen that it throws ConcurrentModificationException exceptions,
+     * on occasions (and why wouldn't it? - our demo app's ActionsScreen has multiple views subscribing and
+     * called concurrently; could it be that not always everything is run in the main thread?).
      * Therefore, we use here {@link ReactFindViewUtil#addViewsListener(OnMultipleViewsFoundListener, Set)},
      * which is too generic but nevertheless allows us to better control when we are to be removed.
      */
-    private static class LongTapsSpy implements OnMultipleViewsFoundListener {
+    private static class LongTapCrasher implements OnMultipleViewsFoundListener {
```
[the remark this answers is not in the document]
Comment (d4vidi) [reply detox/test/android/app/src/main/java/com/example/NativeModule.java]: It does not. It's here because the branch was merged-to from master instead of rebasing it @valentynberehovyi ❤️

[the remark this answers is not in the document]
Comment (d4vidi) [reply detox/test/android/app/src/main/java/com/example/NativeModule.java]: @valentynberehovyi you can still solve this by rebasing the branch. In the future, let's keep this in mind as the reason behind why we always rebase rather than merge 😄 
@rotemmiz #not_nitpickying_anymore

File: detox/android/detox/src/main/java/com/wix/detox/DetoxActionHandlers.kt [hunk 3]
```diff
@@ -110,3 +110,61 @@ class QueryStatusActionHandler(
...
         }
 }
+
+class RecordingStateActionHandler(
+        private val instrumentsManager: DetoxInstrumentsManager,
+        private val wsClient: WebSocketClient
+) : DetoxActionHandler {
+
+    override fun handle(params: String, messageId: Long) {
+        val recordingPath = JSONObject(params).optString("recordingPath", null)
```
Comment (d4vidi) [inline detox/android/detox/src/main/java/com/wix/detox/DetoxActionHandlers.kt]: While this works, the contract by `optString` says the fallback value shouldn't be a `null`. Could you consider using `has()` and `getString()` instead?

File: detox/android/detox/src/main/java/com/wix/detox/DetoxActionHandlers.kt [hunk 4]
```diff
@@ -110,3 +110,61 @@ class QueryStatusActionHandler(
...
+        if (recordingPath != null) {
+            instrumentsManager.startRecordingAtLocalPath(recordingPath)
+        } else {
+            instrumentsManager.stopRecording()
+        }
+
+        wsClient.sendAction("setRecordingStateDone", mapOf<String, Any>(), messageId)
```
Comment (d4vidi) [inline detox/android/detox/src/main/java/com/wix/detox/DetoxActionHandlers.kt]: Can use `emptyMap` instead of `mapOf` here

File: detox/android/detox/src/test/java/com/wix/detox/DetoxActionHandlersSpec.kt [hunk 5]
```diff
@@ -142,5 +142,82 @@ object DetoxActionHandlersSpec: Spek({
                         eq(messageId))
             }
         }
+
+        describe("Recording state actions") {
```
Comment (d4vidi) [inline detox/android/detox/src/test/java/com/wix/detox/DetoxActionHandlersSpec.kt]: This is good, but need to also add a use case checking that done is sent back when path isn't available.

File: detox/android/detox/src/main/java/com/wix/detox/DetoxActionHandlers.kt [hunk 6]
```diff
@@ -110,3 +110,61 @@ class QueryStatusActionHandler(
...
+        private val wsClient: WebSocketClient
+) : DetoxActionHandler {
+
+    override fun handle(params: String, messageId: Long) {
+        val recordingPath = JSONObject(params).optString("recordingPath", null)
+        if (recordingPath != null) {
+            instrumentsManager.startRecordingAtLocalPath(recordingPath)
+        } else {
+            instrumentsManager.stopRecording()
+        }
```
Comment (d4vidi) [inline detox/android/detox/src/main/java/com/wix/detox/DetoxActionHandlers.kt]: While I'm ok with the code, overall, I don't entirely understand the contract here. The name `recordingPath` lacks the meaning that the recording starts; The omission of it lacks the meaning that the recording stops.
I think you guys need to rethink this protocol -- namely, make it more explicit (e.g. record="start"/"stop" alongside recordPath as a parameter when record="start")

File: detox/android/detox/src/test/java/com/wix/detox/DetoxActionHandlersSpec.kt (hunk 5 above)
Comment (d4vidi) [reply detox/android/detox/src/test/java/com/wix/detox/DetoxActionHandlersSpec.kt]: Also, what happens if params is null or malformed (i.e. invalid json)? Better cover that as well (expect as exception)

File: detox/android/detox/src/main/java/com/wix/detox/DetoxActionHandlers.kt [hunk 7]
```diff
@@ -110,3 +110,61 @@ class QueryStatusActionHandler(
             })
         }
 }
+
+class RecordingStateActionHandler(
```
Comment (d4vidi) [inline detox/android/detox/src/main/java/com/wix/detox/DetoxActionHandlers.kt]: Please add an `Instrum` prefix to the calls name so we could categorize it better :)

File: detox/android/detox/src/main/java/com/wix/detox/DetoxActionHandlers.kt [hunk 8]
```diff
@@ -110,3 +110,61 @@ class QueryStatusActionHandler(
...
+            instrumentsManager.startRecordingAtLocalPath(recordingPath)
+        } else {
+            instrumentsManager.stopRecording()
+        }
+
+        wsClient.sendAction("setRecordingStateDone", mapOf<String, Any>(), messageId)
+    }
+}
+
+class EventsActionsHandler(
```
Comment (d4vidi) [inline detox/android/detox/src/main/java/com/wix/detox/DetoxActionHandlers.kt]: `Instrum` prefix plz

File: detox/android/detox/src/main/java/com/wix/detox/DetoxActionHandlers.kt (hunk 4 above)
Comment (valentynberehovyi) [reply detox/android/detox/src/main/java/com/wix/detox/DetoxActionHandlers.kt]: Agree

File: detox/android/detox/src/main/java/com/wix/detox/DetoxActionHandlers.kt (hunk 3 above)
Comment (valentynberehovyi) [reply detox/android/detox/src/main/java/com/wix/detox/DetoxActionHandlers.kt]: has will check for key with any type of value instead of opting string type exactly.

File: detox/android/detox/src/main/java/com/wix/detox/DetoxActionHandlers.kt (hunk 6 above)
Comment (valentynberehovyi) [reply detox/android/detox/src/main/java/com/wix/detox/DetoxActionHandlers.kt]: @d4vidi Protocol is the same as for iOS as well.

File: detox/android/detox/src/test/java/com/wix/detox/DetoxActionHandlersSpec.kt [hunk 9]
```diff
@@ -142,5 +142,82 @@ object DetoxActionHandlersSpec: Spek({
...
+                verify(instrumentsManager).stopRecording()
+            }
+
+            it("should reply with a 'done' ACK on set state finish") {
+                uut().handle(params, messageId)
+                verify(wsClient).sendAction(eq("setRecordingStateDone"), any(), eq(messageId))
+            }
+        }
+
+        describe("Events actions") {
```
Comment (d4vidi) [inline detox/android/detox/src/test/java/com/wix/detox/DetoxActionHandlersSpec.kt]: not bad!

File: detox/android/detox/src/test/java/com/wix/detox/DetoxActionHandlersSpec.kt (hunk 5 above)
Comment (valentynberehovyi) [reply detox/android/detox/src/test/java/com/wix/detox/DetoxActionHandlersSpec.kt]: Agree

File: detox/android/detox/src/test/java/com/wix/detox/DetoxActionHandlersSpec.kt [hunk 10]
```diff
@@ -142,5 +142,82 @@ object DetoxActionHandlersSpec: Spek({
...
+            lateinit var instrumentsManager: DetoxInstrumentsManager
+
+            fun uut() = EventsActionsHandler(instrumentsManager, wsClient)
+
+            beforeEachTest {
+                instrumentsManager = mock()
+            }
+
+            describe("begin interval") {
+                val json = "{\"category\":\"MockCategory\",\"name\":\"MockName\",\"id\":\"MockId\",\"additionalInfo\":\"MockAdditionalInfo\",\"action\":\"begin\"}"
```
Comment (d4vidi) [inline detox/android/detox/src/test/java/com/wix/detox/DetoxActionHandlersSpec.kt]: this is impossible to read :) let's figure out a friendlier solution here. For example, a parser of a multiline JSON string. Or maybe a JSON builder, finally passing to the uut it's parsed (string) form...

Comment (d4vidi) [reply detox/android/detox/src/test/java/com/wix/detox/DetoxActionHandlersSpec.kt]: It's actually gets worse when this content is used in the call to `verify`. Need to come up with a way to predefine and reuse these strings...

File: detox/android/detox/src/test/java/com/wix/detox/DetoxActionHandlersSpec.kt [hunk 11]
```diff
@@ -142,5 +142,82 @@ object DetoxActionHandlersSpec: Spek({
...
+                    verify(instrumentsManager).eventMark(eq("MockCategory"), eq("MockName"), eq("MockId"), eq("MockStatus"), eq("MockAdditionalInfo"))
+                }
+
+                it("should reply with a 'done' ACK") {
+                    uut().handle(json, messageId)
+                    verify(wsClient).sendAction(eq("addActionDone"), any(), eq(messageId))
+                }
+            }
+        }
     }
```
Comment (d4vidi) [inline detox/android/detox/src/test/java/com/wix/detox/DetoxActionHandlersSpec.kt]: Need to also test the 'else'. Please reimplement using a custom, descriptive Detox exception (rather than a RuntimeException); you can add it to the `DetoxErrors` class...

File: detox/android/detox/src/main/java/com/wix/detox/DetoxInstrumentsManager.java [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+            methodEventEndInterval = null;
+            methodEventMark = null;
+        }
+        hasProfiler = methodGetInstanceOfProfiler != null;
+    }
+
+    private final Context context;
+    private Object dtxProfilerInstance;
+
+    DetoxInstrumentsManager(Context context) {
```
Comment (d4vidi) [inline detox/android/detox/src/main/java/com/wix/detox/DetoxInstrumentsManager.java]: Please specify which context type exactly you're expecting here (we have too many in instrumentation testing...)

File: detox/android/detox/src/test/java/com/wix/detox/DetoxActionHandlersSpec.kt (hunk 11 above)
Comment (valentynberehovyi) [reply detox/android/detox/src/test/java/com/wix/detox/DetoxActionHandlersSpec.kt]: Good catch! Agree.

File: detox/android/detox/src/main/java/com/wix/detox/DetoxManager.java [hunk 13]
```diff
@@ -145,5 +145,9 @@ public Unit invoke() {
                 return null;
             }
         }));
+
+        final DetoxInstrumentsManager instrumentsManager = new DetoxInstrumentsManager(reactNativeHostHolder);
+        actionHandlers.put("setRecordingState", new RecordingStateActionHandler(instrumentsManager, wsClient));
+        actionHandlers.put("addEvent", new EventsActionsHandler(instrumentsManager, wsClient));
```
Comment (d4vidi) [inline detox/android/detox/src/main/java/com/wix/detox/DetoxManager.java]: Could we change this so these will only be added conditionally (i.e. if indeed instrumentation exists)?

File: detox/android/detox/src/main/java/com/wix/detox/DetoxInstrumentsManager.java [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+                    boolean.class//recordReactNativeTimersAsEvents
+            );
+            methodGetInstanceOfProfiler = profilerClass.getDeclaredMethod("getInstance", Context.class);
+            methodStartRecording = profilerClass.getDeclaredMethod("startProfiling", Context.class, configurationClass);
+            methodStopRecording = profilerClass.getDeclaredMethod("stopProfiling");
+            methodEventBeginInterval = profilerClass.getDeclaredMethod("eventBeginInterval", String.class, String.class, String.class, String.class, String.class);
+            methodEventEndInterval = profilerClass.getDeclaredMethod("eventEndInterval", String.class, String.class, String.class);
+            methodEventMark = profilerClass.getDeclaredMethod("eventMark", String.class, String.class, String.class, String.class, String.class);
+        } catch (Exception e) {
+            Log.i("DetoxInstrumentsManager", "Instruments not found", e);
```
Comment (d4vidi) [inline detox/android/detox/src/main/java/com/wix/detox/DetoxInstrumentsManager.java]: I'm not happy with detox starting up in standard mode with an exception on the device. Let's not `static` this and run this conditionally, and only when the proper time comes during detox init....

File: detox/android/detox/src/main/java/com/wix/detox/DetoxInstrumentsManager.java [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+                    long.class,//samplingIntervalMillis
+                    File.class,//recordingFile
+                    boolean.class//recordReactNativeTimersAsEvents
+            );
+            methodGetInstanceOfProfiler = profilerClass.getDeclaredMethod("getInstance", Context.class);
+            methodStartRecording = profilerClass.getDeclaredMethod("startProfiling", Context.class, configurationClass);
+            methodStopRecording = profilerClass.getDeclaredMethod("stopProfiling");
+            methodEventBeginInterval = profilerClass.getDeclaredMethod("eventBeginInterval", String.class, String.class, String.class, String.class, String.class);
+            methodEventEndInterval = profilerClass.getDeclaredMethod("eventEndInterval", String.class, String.class, String.class);
+            methodEventMark = profilerClass.getDeclaredMethod("eventMark", String.class, String.class, String.class, String.class, String.class);
```
Comment (d4vidi) [inline detox/android/detox/src/main/java/com/wix/detox/DetoxInstrumentsManager.java]: I can't accept this... could you please abstract this away in a delegate (for example, a `DetoxProfilerReflected` class?)

File: detox/android/detox/src/main/java/com/wix/detox/DetoxInstrumentsManager.java [hunk 16]
```diff
@@ -0,0 +1,141 @@
+package com.wix.detox;
+
+import android.content.Context;
+import android.util.Log;
+
+import java.io.File;
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Method;
+
+public class DetoxInstrumentsManager {
```
Comment (d4vidi) [inline detox/android/detox/src/main/java/com/wix/detox/DetoxInstrumentsManager.java]: What about `DetoxProfilingProxy`? 🤷‍♂

File: detox/android/detox/src/main/java/com/wix/detox/DetoxInstrumentsManager.java [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+    }
+
+    private Object obtainProfiler() {
+        try {
+            return methodGetInstanceOfProfiler.invoke(null, context);
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+}
```
Comment (d4vidi) [inline detox/android/detox/src/main/java/com/wix/detox/DetoxInstrumentsManager.java]: As in some of my previous comments, let's move some of this to a delegate class (e.g. `DetoxProfilerReflected`) and un-static stuff and then add some unit tests, plzzzz

File: detox/android/detox/src/main/java/com/wix/detox/DetoxActionHandlers.kt (hunk 6 above)
Comment (noomorph) [reply detox/android/detox/src/main/java/com/wix/detox/DetoxActionHandlers.kt]: @d4vidi, Valentyn is right. That is what was implemented for iOS.

File: detox/src/artifacts/instruments/InstrumentsArtifactRecording.js [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+  }
+
+  async doStop() {
+    if (this._isClientConnected()) {
+      await this._client.stopInstrumentsRecording();
+    }
+  }
+
+  _isClientConnected() {
+    return this._client.isConnected && !this._client.pandingAppCrash;
```
Comment (noomorph) [inline detox/src/artifacts/instruments/InstrumentsArtifactRecording.js]: @valentynberehovyi, we'll need to merge the latest changes from master to here.

File: detox/src/artifacts/instruments/android/AndroidInstrumentsRecording.js [hunk 19]
```diff
@@ -0,0 +1,22 @@
...
+class AndroidInstrumentsRecording extends InstrumentsArtifactRecording {
+  constructor({ adb, client, deviceId, temporaryRecordingPath }) {
+    super({ client, temporaryRecordingPath });
+    this.adb = adb;
+    this.deviceId = deviceId;
+  }
+
+  async doSave(artifactPath) {
+    await super.doSave(artifactPath);
+    await this.adb.pull(this.deviceId, this.temporaryRecordingPath, artifactPath);
```
Comment (noomorph) [inline detox/src/artifacts/instruments/android/AndroidInstrumentsRecording.js]: Just musing, what will we do if we don't find a file there... ? If actual profiler code won't be there, I mean. @valentynberehovyi

Comment (noomorph) [reply detox/src/artifacts/instruments/android/AndroidInstrumentsRecording.js]: Should we hint as well like below... ?
```
SimulatorInstrumentsRecording.hintAboutDetoxInstruments = _.once(() => {
   log.warn(`Make sure either:
 1. You have installed Detox Instruments:
    https://github.com/wix/DetoxInstruments#installation 
 2. You have integrated Detox Instruments in your app:
    https://github.com/wix/DetoxInstruments/blob/master/Documentation/XcodeIntegrationGuide.md 
 3. You have set the environment variable with your custom Detox Instruments location:
    export DETOX_INSTRUMENTS_PATH="/path/to/Detox Instruments.app"`);
 });
```

Comment (valentynberehovyi) [reply detox/src/artifacts/instruments/android/AndroidInstrumentsRecording.js]: We'll add this later. For now recordings will not start when instruments is not installed.

File: detox/android/detox/build.gradle [hunk 20]
```diff
@@ -93,6 +93,16 @@ dependencies {
...
+    buildscript {
+        repositories {
+            jcenter()
+        }
+
+        dependencies {
+            classpath 'de.mannodermaus.gradle.plugins:android-junit5:1.5.2.0'
+        }
+    }
+
```
Comment (d4vidi) [inline detox/android/detox/build.gradle]: @valentynberehovyi why is this needed? it's already in the root build.gradle, were you trying to move it?

File: detox/android/detox/src/main/java/com/wix/detox/DetoxActionHandlers.kt (hunk 6 above)
Comment (d4vidi) [reply detox/android/detox/src/main/java/com/wix/detox/DetoxActionHandlers.kt]: Would you consider changing this?

File: detox/android/detox/build.gradle (hunk 20 above)
Comment (valentynberehovyi) [reply detox/android/detox/build.gradle]: In case of using detox as android project dependency classpath from root project is not available.

File: detox/android/detox/src/main/java/com/wix/detox/DetoxInstrumentsManager.java (hunk 12 above)
Comment (d4vidi) [reply detox/android/detox/src/main/java/com/wix/detox/DetoxInstrumentsManager.java]: This is still applicable

File: detox/android/detox/build.gradle (hunk 20 above)
Comment (d4vidi) [reply detox/android/detox/build.gradle]: Could you elaborate? what wasn't working specifically?

File: detox/android/detox/src/main/java/com/wix/detox/instruments/DetoxInstrumentsManager.java [hunk 21]
```diff
@@ -0,0 +1,79 @@
...
+            String eventStatus,
+            String additionalInfo
+    ) {
+        if (recording != null) {
+            recording.eventMark(category, name, id, eventStatus, additionalInfo);
+        }
+    }
+
+    public static boolean supports() {
+        return getDefaultInstruments().installed();
```
Comment (d4vidi) [inline detox/android/detox/src/main/java/com/wix/detox/instruments/DetoxInstrumentsManager.java]: All initializations made in order to answer this question are done static-ly, so is it really necessary to have an instance here? It seems a bit redundant and also I found it a bit confusing tbh 🤷‍♂

File: detox/android/detox/src/main/java/com/wix/detox/instruments/reflected/InstrumentsRecordingReflected.java [hunk 22]
```diff
@@ -0,0 +1,109 @@
+package com.wix.detox.instruments.reflected;
+
+import com.wix.detox.instruments.DetoxInstrumentsException;
+import com.wix.detox.instruments.InstrumentsRecording;
+
+import java.lang.reflect.Method;
+
+public class InstrumentsRecordingReflected implements InstrumentsRecording {
```
Comment (d4vidi) [inline detox/android/detox/src/main/java/com/wix/detox/instruments/reflected/InstrumentsRecordingReflected.java]: Love this, so clean...

File: detox/android/detox/src/test/java/com/wix/detox/instruments/DetoxInstrumentsManagerSpec.kt [hunk 23]
```diff
@@ -0,0 +1,130 @@
...
+import com.wix.invoke.MethodInvocation
+import org.json.JSONObject
+import org.spekframework.spek2.Spek
+import org.spekframework.spek2.style.specification.describe
+import java.io.File
+import java.lang.reflect.InvocationTargetException
+import java.util.*
+import java.util.concurrent.Executors
+
+object DetoxInstrumentsManagerSpec : Spek({
```
Comment (d4vidi) [inline detox/android/detox/src/test/java/com/wix/detox/instruments/DetoxInstrumentsManagerSpec.kt]: boom

File: detox/android/detox/build.gradle (hunk 20 above)
Comment (valentynberehovyi) [reply detox/android/detox/build.gradle]: Spek tests were not working in case of using detox as android project dependency.

File: detox/android/detox/src/main/java/com/wix/detox/DetoxActionHandlers.kt (hunk 6 above)
Comment (valentynberehovyi) [reply detox/android/detox/src/main/java/com/wix/detox/DetoxActionHandlers.kt]: Not for this iteration. As then we'll need to change the iOS part as well.

File: detox/android/detox/src/main/java/com/wix/detox/instruments/DetoxInstrumentsManager.java (hunk 21 above)
Comment (valentynberehovyi) [reply detox/android/detox/src/main/java/com/wix/detox/instruments/DetoxInstrumentsManager.java]: Static-ly initialised only *Reflected part, which is only one of possible instruments implementations.

File: detox/android/detox/build.gradle (hunk 20 above)
Comment (rotemmiz) [reply detox/android/detox/build.gradle]: @valentynberehovyi can you explain exactly what failed (what gradle task/command) for you that required you to add this?

Comment (valentynberehovyi) [reply detox/android/detox/build.gradle]: Mine test app uses detox as project dependency. Also mine app set ext.isOfficialDetoxApp=true to develop test spek as well.
Without classpath in buildscript dependency gradle not able to resolve plugin **'de.mannodermaus.android-junit5'** . I'll move it to mine root project also for now.

File: detox/src/artifacts/instruments/InstrumentsArtifactRecording.js (hunk 18 above)
Comment (noomorph) [reply detox/src/artifacts/instruments/InstrumentsArtifactRecording.js]: @valentynberehovyi , what is the status here?
